src/chainstate/stacks/index/trie.rs

```diff
@@ -0,0 +1,2053 @@
+            trace!("Fix up root node so it mixes in its ancestor hashes");
+            let (node, cur_hash) = storage.read_nodetype(&child_ptr)?;
+            if node.is_node256() {
+                let root_disk_ptr = storage.root_ptr();
+                let root_ptr = TriePtr::new(TrieNodeID::Node256, 0, root_disk_ptr as u32);
+                if child_ptr != root_ptr {
+                    return Err(Error::CorruptionError("Only ptr is not the root".to_string()));
+                }
+                
+                let mut hash_buf = Vec::with_capacity(TRIEHASH_ENCODED_SIZE * 256);
+                Trie::get_children_hashes_bytes(storage, &node, &mut hash_buf)?;
+
+                let h = 
+                    if update_skiplist {
+                        trace!("Update root skiplist");
+                        Trie::get_trie_root_hash(storage, &get_nodetype_hash_bytes(&node, &hash_buf))?
+                    }
+                    else {
+                        trace!("Not updating root skiplist");
+                        get_nodetype_hash_bytes(&node, &hash_buf)
+                    };
+
+                // for debug purposes
+                if is_trace() {
+                    let node_hash = get_nodetype_hash_bytes(&node, &hash_buf);
+                    let hs = Trie::get_trie_root_ancestor_hashes(storage, &node_hash)?;
+                    trace!("update_root_hash: Updated {:?} with {:?} from {:?} to {:?} + {:?} = {:?} (fixed root)", &node, &child_ptr, &cur_hash, &node_hash, &hs[1..].to_vec(), &h);
+                }
+
+                storage.write_nodetype(child_ptr.ptr(), &node, h)?;
+            }
+            else {
+                return Err(Error::CorruptionError("Only ptr was not a node256".to_string()));
+            }
+        }
+        else {
+            while ptrs.len() > 0 {
+                let ptr = match ptrs.pop() {
+                    Some(p) => p,
+                    None => panic!("Out of ptrs")
+                };
+                if is_backptr(ptr.id()) {
+                    // this node was not altered, but instead queued to the cursor as part of walking a
+                    // backptr skiplist.  Do nothing.
+                    continue;
+                }
+
+                let (mut node, cur_hash) = storage.read_nodetype(&ptr)?;
+                assert!(!node.is_leaf());
+
+                // this child_ptr _must_ be in the node.
+                let updated = node.replace(&child_ptr);
+                if !updated {
+                    trace!("FAILED TO UPDATE {:?} WITH {:?}: {:?}", &node, &child_ptr, cursor);
+                    assert!(updated);
+                }
+
+                let mut hash_buf = Vec::with_capacity(TRIEHASH_ENCODED_SIZE * 256);
+                Trie::get_children_hashes_bytes(storage, &node, &mut hash_buf)?;
+
+                if !node.is_node256() {
+                    let h = get_nodetype_hash_bytes(&node, &hash_buf);
```

Comment:
It seems like there's a few ways that hashes will get recomputed twice --  `insert_leaf` will compute hashes, and then it'll get computed again here. It'd be preferable if there was a single function that was responsible for performing the recomputation.

Reply:
Yeah, I think you're right.  I think the way to avoid recalculation would be to tag ptrs in the TrieCursor as already having the "latest" hash, so this method can just skip updating them.

Reply:
Alternatively, (per your comment below), the insertion methods should _not_ calculate hashes, and instead, _all_ hashes should be calculated in the `update_root_hash()` method.

Reply:
So I took a stab at this.  Making the `update_root_hash()` method exclusively responsible for calculating all hashes is a bad idea, for two reasons:

* Doing so is at odds with point (3) in your review below.  The methods that alter the Trie structure will need to update the TrieCursor with information on the node(s) whose hashes are now invalid.

* Trie update methods like `splice_leaf()` and `promote_leaf_to_node4()` touch two or more branches, meaning that the branch information would need to be conveyed via the TrieCursor to `update_root_hash()`.

I think the most all-around reasonable thing to do in order to ensure that all modified nodes' and their parents' hashes are recalculated exactly one is to design the Trie mutation methods to guarantee that the _highest_ node they touch will have the correct hash when the method returns (which implies that they will ensure that all nodes beneath it have the correct hashes).   
 I think this is appropriate because these methods _already know_ which nodes' hashes need to be recalculated.  This is more or less how they work today, but there are two instances where a node's hash is calculated twice (which I have fixed).  Then, `update_root_hash()` only needs to know the highest node modified in order to update the root hash.

Per point (4) below, I think that this gets us the an acceptable correctness test -- an auditor would need to confirm that:

* All mutation methods render the sub-Trie beneath the highest-modified node well-formed (i.e. all nodes have the right hashes).  This is trivial for leaf insert/replace, and relatively straightforward to confirm for node-promotion and path-splicing.
* `update_root_hash()` re-calculates all node hashes starting from the parent of the root of the modified sub-Trie.

What do you think?

Reply:
Yep -- that makes sense to me.